accepted/0000-selective-versions-resolutions.md

```diff
@@ -0,0 +1,423 @@
+seen it happen many times with maven, which is quite complex but complete in
+its dependency management. Users would get confused and it can take time to
+understand the implications of manipulation the `resolutions` field .
+
+# Alternatives
+
+## Global nested dependencies resolution
+
+Starting from an example, this solution would take the following form in the
+`package.json` file:
+```json
+  "devDependencies": {
+    "@angular/cli": "1.0.3",
+    "typescript": "2.3.2"
+  },
+  "resolutions": {
+    "typescript": "2.0.2"
+  }
+```
+
+yarn would use `typescript@2.0.2` for the whole project and that's all.
+The same kind of consideration (outside of the glob pattern thing) should be
+followed that in the RFC selected solution.
+
+## Mapping version specifications
+
+This is a kind of simplified solution to the "out-of-scope scenario" presented
+in the motivations section above (it maps versions but not dependency names).
+
+It was proposed in this
+[comment](https://github.com/yarnpkg/yarn/issues/2763#issuecomment-301896274).
+
+It is similar to the previous alternative but with a version specification
+in the package designation. This would take this form in the `package.json`:
+```json
+"devDependencies": {
+    "@angular/cli": "1.0.3",
+    "typescript": "2.2.2",
+    "more dependencies..."
+  },
+  "resolutions": {
+    "typescript@>=2.0.0 <2.3.0": "typescript@2.3.2"
+  }
+```
+
+yarn would then replace matching version specifications with the user's one.
+For example a dependency normally resolved to `typescript@2.2.2` would be
+resolved in practice to `typescript@2.3.2`.
+
+## Mapping version specifications as well as packages name
+
+Same as the tow above but with a different name on the right-hand side of the
+*resolution*:
+```json
+"devDependencies": {
+    "@angular/cli": "1.0.3",
+    "typescript": "2.2.2"
+  },
+  "resolutions": {
+    "typescript@>=2.0.0 <2.3.0": "my-typescript-fork@2.3.2"
+  }
+```
+
+or even:
+```json
+"devDependencies": {
+    "@angular/cli": "1.0.3",
+    "typescript": "2.2.2"
+  },
+  "resolutions": {
+    "typescript": "my-typescript-fork",
+  }
+```
+
+and the version specification would be conserved.
+
+# Unresolved questions
+
+## Package designation and single `*`
+
+I am not totally convinced that using a single `*` in a package designation
+is a good idea. Mostly because it introduces much for uncertainty to what it is
+going to be applied. In particular, I'm thinking about defining a *resolution*
+like that and latter adding a dependency which has a dependency that matches
+but is not intended to. I feel like it adds a lot of mental weight for the
+user to manipulate single `*`.
+
+## `--flat` option
+
+When used with the `install` command, the behaviour of `--flat` with respect
+to this RFC is clear: it will populate the `resolutions` for all packages with
+`**/dependency` as package designations.
+
+But when used with the `add` command, should it apply only to the nested

```

Comment:
no, it is about what would be the behaviour of `yarn add --flat left-pad`:
1. should it flatten (i.e. write *resolutions* for) only the dependencies of `left-pad` without flattening the rest of the nested dependencies of the project (except for those that are common to `left-pad`)
2. should it flatten (i.e., write *resolutions* for) the whole project (as it currently does).